src/audio/up_down_mixer/up_down_mixer.c

```diff
@@ -0,0 +1,532 @@
+		ret = init_mix(cd, &up_down_mixer->base_cfg.audio_fmt, up_down_mixer->out_channel_config, NULL);
+		break;
+	case CUSTOM_COEFFICIENTS_WITH_CHANNEL_MAP:
+		cd->out_channel_map = up_down_mixer->channel_map;
+		ret = init_mix(cd, &up_down_mixer->base_cfg.audio_fmt, up_down_mixer->out_channel_config,
+			       up_down_mixer->coefficients);
+		break;
+	default:
+		comp_err(dev, "init_up_down_mixer(): unsupported coefficient type");
+		up_down_mixer_free(dev);
+		return -EINVAL;
+	}
+
+	if (!cd->mix_routine) {
+		up_down_mixer_free(dev);
+		comp_err(dev, "init_up_down_mixer(): mix routine unitialized.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static struct comp_dev *up_down_mixer_new(const struct comp_driver *drv,
+					  struct comp_ipc_config *config,
+					  void *spec)
+{
+	struct comp_dev *dev;
+	int ret;
+
+	comp_cl_info(&comp_up_down_mixer, "up_down_mixer_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	ret = init_up_down_mixer(dev, config, spec);
+	if (ret < 0) {
+		comp_free(dev);
+		return NULL;
+	}
+	dev->state = COMP_STATE_READY;
+
+	return dev;
+}
+
+static int up_down_mixer_prepare(struct comp_dev *dev)
+{
+	int ret;
+
+	if (dev->state == COMP_STATE_ACTIVE) {
+		comp_info(dev, "up_down_mixer_prepare(): Component is in active state.");
+		return 0;
+	}
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
+		return PPL_STATUS_PATH_STOP;
+
+	return 0;
+}
+
+static int up_down_mixer_reset(struct comp_dev *dev)
+{
+	return 0;
+}
+
+static int up_down_mixer_copy(struct comp_dev *dev)
+{
+	struct comp_buffer *source;
+	struct comp_buffer *sink;
+	uint32_t source_bytes, sink_bytes;
+	uint32_t mix_frames;
+	uint32_t buff_frag = 0;
+	int32_t *destp;
+	unsigned int i, channel;
+	struct up_down_mixer_data *cd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "up_down_mixer_copy()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 sink_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       source_list);
+
+	mix_frames = MIN(audio_stream_get_avail_frames(&source->stream),
+			 audio_stream_get_free_frames(&sink->stream));
+
+	source_bytes = mix_frames * audio_stream_frame_bytes(&source->stream);
+	sink_bytes = mix_frames * audio_stream_frame_bytes(&sink->stream);
+
+	if (!source_bytes)
+		return 0;
+
+	buffer_stream_invalidate(source, source_bytes);
+
+	for (i = 0; i < mix_frames; i++) {
+		for (channel = 0; channel < source->stream.channels; channel++) {
+			cd->buf_in[buff_frag] = *(int32_t *)audio_stream_read_frag_s32(&source->stream, buff_frag);
+
+			buff_frag++;
+		}
+	}
+
+	cd->mix_routine(cd, (uint8_t *)cd->buf_in, source_bytes, (uint8_t *)cd->buf_out);
+
+	buff_frag = 0;
+	for (i = 0; i < mix_frames; i++) {
+		for (channel = 0; channel < sink->stream.channels; channel++) {
+			destp = audio_stream_write_frag_s32(&sink->stream, buff_frag);
```

Comment:
Yep, we should avoid these. I'm now collecting list of components' status to see which components need update before we can remove the too slow functions. The current fixing status is in #4967. You can look from there how to avoid the sample by sample read/write frag operations.

Reply:
Removed write_frag, now we have audio_stream_copy_to_linear() and audio_stream_copy_from_linear(), please review